File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/util/PrefUtil.java
package de.deutschebahn.bahnhoflive.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import androidx.annotation.NonNull;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import de.deutschebahn.bahnhoflive.backend.db.fasta2.model.FacilityStatus;

public class PrefUtil {

	private static final String SAVED_FACILITIES = "38";
	private static final String FACILITIES_PUSH_ACTIVE = "39";


	private static SharedPreferences getPrefs(Context context) {
		return PreferenceManager.getDefaultSharedPreferences(context);
	}


	/*
	 *
	 *
	 * CACHING
	 *
	 *
	 */


	public static void storeAlarmKey(@NonNull final String key, @NonNull Context context) {

		final String alarmKey = String.format("alarm_%s", key);

		getPrefs(context)
				.edit()
				.putString(alarmKey, key)
				.commit();

	}

	public static void cleanAlarmKey(@NonNull  String key, @NonNull  Context context) {
		final String alarmKey = String.format("alarm_%s", key);
		getPrefs(context).edit().remove(alarmKey).commit();
	}

	public static boolean hasAlarmSet(@NonNull final String key, @NonNull Context context) {
		final String alarmKey = String.format("alarm_%s", key);
		return getPrefs(context).contains(alarmKey);
	}


	public static List<FacilityStatus> getSavedFacilities(Context context) {
		String savedString = getPrefs(context).getString(SAVED_FACILITIES, "[]");
		return FacilityStatus.fromString(savedString);
	}



	public static boolean getFacilitySubscribed(Context context, int equipmentNumber) {
		List<FacilityStatus> facilities = getSavedFacilities(context);
		for (FacilityStatus f : facilities) {
			if (f.getEquipmentNumber() == equipmentNumber) {
				return f.isSubscribed();
			}
		}
		return false;
	}

	public static void setFacilitySubcribed(Context context, FacilityStatus facility, boolean isChecked) {
		List<FacilityStatus> savedFacilities = getSavedFacilities(context);
		for (FacilityStatus fs : savedFacilities) {
			if (fs.getEquipmentNumber() == facility.getEquipmentNumber()) {
				//modifiy saved facility
				fs.setSubscribed(isChecked);
				storeSavedFacilities(context, savedFacilities);
				return;
			}
		}

		//add facility
		facility.setSubscribed(isChecked);
		savedFacilities.add(facility);
		storeSavedFacilities(context, savedFacilities);
	}

	public static void storeSavedFacilities(Context context, List<FacilityStatus> facilities) {
		//sort by station name and then by description
		Collections.sort(facilities, new Comparator<FacilityStatus>() {
			@Override
			public int compare(FacilityStatus f1, FacilityStatus f2) {
				int res = f1.getStationName().compareTo( f2.getStationName() );
				if(res == 0){
					return f1.getDescription().compareTo( f2.getDescription() );
				}
				return res;
			}
		});
		
		getPrefs(context)
				.edit()
				.putString(SAVED_FACILITIES, FacilityStatus.toString(facilities))
				.commit();
	}

	public static void removeSavedFacilityStatus(Context context, FacilityStatus facilityStatus) {
		List<FacilityStatus> savedFacilities = getSavedFacilities(context);
		for (FacilityStatus fs : savedFacilities) {
			if (fs.getEquipmentNumber() == facilityStatus.getEquipmentNumber()) {
				savedFacilities.remove(fs);
				storeSavedFacilities(context, savedFacilities);
				return;
			}
		}
	}

	public static boolean getFacilityPushActivated(Context context) {
		return getPrefs(context).getBoolean(FACILITIES_PUSH_ACTIVE, true);
	}

	public static void setFacilityGlobalPushActivated(Context context, boolean active) {
		getPrefs(context)
				.edit()
				.putBoolean(FACILITIES_PUSH_ACTIVE,active)
				.commit();
	}

}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/util/PrefUtil.kt
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package de.deutschebahn.bahnhoflive.util

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import de.deutschebahn.bahnhoflive.backend.db.fasta2.model.FacilityStatus
import java.util.Collections

object PrefUtil {
    private const val SAVED_FACILITIES = "38"
    private const val FACILITIES_PUSH_ACTIVE = "39"
    const val FORMAT_ALARM = "alarm_%s"

    private fun getPrefs(context: Context): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(context)
    }

    /*
     *
     *
     * CACHING
     *
     *
     */
    @JvmStatic
    fun storeAlarmKey(key: String, context: Context) {
        val alarmKey = String.format(FORMAT_ALARM, key)
        getPrefs(context)
            .edit()
            .putString(alarmKey, key)
            .commit()
    }

    @JvmStatic
    fun cleanAlarmKey(key: String, context: Context) {
        val alarmKey = String.format(FORMAT_ALARM, key)
        getPrefs(context).edit().remove(alarmKey).commit()
    }

    @JvmStatic
    fun hasAlarmSet(key: String, context: Context): Boolean {
        val alarmKey = String.format(FORMAT_ALARM, key)
        return getPrefs(context).contains(alarmKey)
    }

    @JvmStatic
    fun getSavedFacilities(context: Context): MutableList<FacilityStatus> {
        val savedString = getPrefs(context).getString(SAVED_FACILITIES, "[]")
        return FacilityStatus.fromString(savedString)
    }

    fun getFacilityBookmarked(context: Context, equipmentNumber: Int): Boolean {
        val facilities: List<FacilityStatus> = getSavedFacilities(context)
        for (f in facilities) {
            if (f.equipmentNumber == equipmentNumber) {
                return f.isBookmarked
            }
        }
        return false
    }

    fun setFacilityBookmarked(context: Context, facility: FacilityStatus, isBookmarked: Boolean) {
        val savedFacilities = getSavedFacilities(context)
        for (fs in savedFacilities) {
            if (fs.equipmentNumber == facility.equipmentNumber) {
                //modifiy saved facility
                fs.isBookmarked = isBookmarked
                storeSavedFacilities(context, savedFacilities)
                return
            }
        }

        //add facility
        facility.isBookmarked = isBookmarked
        savedFacilities.add(facility)
        storeSavedFacilities(context, savedFacilities)
    }

    @JvmStatic
    fun storeSavedFacilities(context: Context, facilities: List<FacilityStatus>?) {
        //sort by station name and then by description
        Collections.sort(facilities) { f1, f2 ->
            val res = f1.stationName.compareTo(f2.stationName)
            if (res == 0) {
                f1.description.compareTo(f2.description)
            } else res
        }
        getPrefs(context)
            .edit()
            .putString(SAVED_FACILITIES, FacilityStatus.toString(facilities))
            .commit()
    }

    fun removeSavedFacilityStatus(context: Context, facilityStatus: FacilityStatus) {
        val savedFacilities = getSavedFacilities(context)
        for (fs in savedFacilities) {
            if (fs.equipmentNumber == facilityStatus.equipmentNumber) {
                savedFacilities.remove(fs)
                storeSavedFacilities(context, savedFacilities)
                return
            }
        }
    }


    fun getFacilityPushEnabled(context: Context, facility: FacilityStatus): Boolean {

        val savedFacilitie =
            getSavedFacilities(context).firstOrNull { it.equipmentNumber == facility.equipmentNumber }

        if(savedFacilitie!=null)  {
            return  savedFacilitie.canReceivePushMessages()
        }
        return false
    }

    fun getFacilityPushEnabled(context: Context, equipmentNumber: Int): Boolean {

        val savedFacilitie =
            getSavedFacilities(context).firstOrNull { it.equipmentNumber == equipmentNumber }

        if(savedFacilitie!=null)  {
            return  savedFacilitie.canReceivePushMessages()
        }
        return false
    }


    fun setFacilityPushEnabled(context: Context,  facility: FacilityStatus, enabled: Boolean) {

        val savedFacilities = getSavedFacilities(context)
        for (fs in savedFacilities) {
            if (fs.equipmentNumber == facility.equipmentNumber) {
                //modifiy saved facility
                fs.enableReceivePushMessages(enabled)
                storeSavedFacilities(context, savedFacilities)
                return
            }
        }

        //add facility
        facility.isBookmarked = enabled
        savedFacilities.add(facility)
        storeSavedFacilities(context, savedFacilities)
    }
}